doc: use MathJax for HTML documentation

japm48 opened the pull request:
Fixes #2707.

A full LaTeX installation is no longer required for generating the HTML docs.

Apparently this was previously broken (?), e.g.: https://www.gnuradio.org/doc/doxygen/classgr_1_1fec_1_1code_1_1ldpc__gen__mtrx__encoder.html

Note that this adds a newer dependency but replaces the former hefty LaTeX dependency (which was undocumented?).

777arc commented:
So, looking at our [list of dependencies for Ubuntu 18](https://wiki.gnuradio.org/index.php/UbuntuInstall#Bionic_Beaver_.2818.04.29_through_Eoan_Ermine_.2819.10.29), I don't even see the LaTeX one, and sure enough when I look at the doxygen of builds I've created, the equations are not rendered into LaTeX, they just show up between $'s.  

When I test this new PR, Doxygen becomes disabled because MATHJAX2 isn't found.  I did an apt-get install libjs-mathjax, and that made Doxygen enabled for the build.  So we'll have to add libjs-mathjax to the list, which I'm fine with.  But if it's possible can you **remove the requirement for mathjax**, so that it still builds the doxygen without it, like when LaTeX wasn't found? 

The resulting docs look good!  My build worked, once I disabled GR_DIGITAL which was failing because of some SWIG reason I figured was unrelated to your PR.

japm48 commented:
@777arc

Ok, I just tried removing the hard dependency in the CMake file.

But then I see two options:

- Leave it as it is now: in this case, if MathJax is found at compile time, it gets hardcoded into the generated HTML files (in a `<script>` tag). If it is not found, the default remote MathJax.js is used instead; the problem is that this is strongly discouraged ([source](http://www.doxygen.nl/manual/config.html#cfg_mathjax_relpath)), possibly to avoid saturating their servers.
 - Another option is to disable MathJax if not found at compile time (`USE_MATHJAX=NO`), so it will automatically try with the previous LaTeX method.

What do you suggest that should happen?

Apart from that, I believe this dependency should still appear in the documentation (i.e. in `build_guide.dox.in`).

777arc commented:
It's a tough one because I'm pretty sure 99.9% of people aren't ever looking at the docs they build, if they do look at docs its the ones hosted online.  And we didn't even have LaTeX listed as a dependency in several spots.  I personally don't care which way you go, as long as it doesn't fully disable doxygen when MathJax is missing, simply because once we do this change there will be tons of people installing it without MathJax, and I don't want folks to get thrown off if they see Doxygen disabled all of a sudden, even if they aren't referencing the docs.  

We should get @noc0lour to thumbs up just because in the past he ran the doxygen hosting, although I've been updating it myself lately.

japm48 commented:
In the end, I came up with this: emit a warning when compiling docs if mathjax is not found.

Also, I realized that you cannot deploy a web that points to `/usr/share/...` so you will have to host a copy of the folder that contains `MathJax.js` (or use a VirtualHost or however this is done these days :)).

For that you can use `-DMATHJAX2_FORCED_ROOT=...url...` when invoking cmake (relative URLs are also fine).
You can test this with e.g. `-DMATHJAX2_FORCED_ROOT=https://cdn.jsdelivr.net/npm/mathjax@2`

777arc reviewed (approved):
I think we should get this merged in, after 1 more person tests that it doesn't do anything weird to a normal build (e.g. it should still build doxygen when mathjax isn't installed), and then we need to make sure that our process for building and uploading the doxygen manual to our website includes mathjax and gets the LaTeX to compile properly.  So I'll reach out on Matrix and try to push things along.

drmpeg reviewed (approved):
I've tested this PR. It works fine with/without Latex and/or Mathjax.

drmpeg commented:
I'm seeing an issue with Mathjax extensions. From the Web console:
```
Loading failed for the <script> with source “file:///usr/share/javascript/mathjax/extensions/TeX.js?V=2.7.3”.
```
Seems like some sort of localization thing since Tex.js is in:
```
/usr/share/javascript/mathjax/localization/en
```

drmpeg commented:
Hold the phone. OOT's are not finding Mathjax properly.

drmpeg commented on gr-utils/modtool/templates/gr-newmod/docs/CMakeLists.txt:
```diff
@@ -10,6 +10,7 @@
 # Setup dependencies
 ########################################################################
 find_package(Doxygen)
+find_package(MathJax)
```
This line needs to be removed. It will look for FindMathJax2.cmake which is not present in the OOT.

To find the package in the OOT, add the line:
```find_dependency(MathJax2)```
in cmake/Modules/GnuRadioConfig.cmake.in

drmpeg commented on gr-utils/modtool/templates/gr-newmod/docs/CMakeLists.txt:
```diff
@@ -10,6 +10,7 @@
 # Setup dependencies
 ########################################################################
 find_package(Doxygen)
+find_package(MathJax)
```
This line needs to be removed. It will try to find FIndMathJax2.cmake in the OOT, which is not present.

Unfortunately, I don't have an easy fix for this yet.

777arc replied:
Are there even OOTs out there that make use of LaTeX in their Doxygen manual?  Important thing is to fix it within GR, having it work within OOTs would be nice but it shouldn't block the GR side imo.

drmpeg replied:
That would be one way to solve the issue.

japm48 commented:
I just changed this:

* Fixed `TeX.js` issue, this was my fault (Copy-Paste bug), and nothing to do with localization.

* By default OOT modules will not support MathJax (`enable_mathjax=NO`); this is probably the best option as it is pretty easy to manually add the cmake changes if desired. Note that it was also the case for LaTeX (`enable_latex_docs=NO`).

drmpeg commented on docs/doxygen/Doxyfile.in:
```diff
@@ -1328,12 +1321,12 @@ MATHJAX_FORMAT         = HTML-CSS
 # names that should be enabled during MathJax rendering.
 
-MATHJAX_EXTENSIONS     =
+MATHJAX_EXTENSIONS     = TeX/AMSmath TeX/AMSsymbols TeX/AMSmath
```
Is it necessary to have Tex/AMSmath listed twice?

japm48 replied:
No, fixed!

drmpeg commented:
Looks good.

noc0lour commented:
Ideally we have the doc building tied in the CI to automate building developers & stable docs, other than that anything really flies with me. That's something for the LaTeX formula necromancer @marcusmueller to decide.

mormj commented:
ping @marcusmueller

marcusmueller commented on docs/CMakeLists.txt:
```diff
@@ -9,12 +9,14 @@
 ########################################################################
 include(GrComponent)
-GR_REGISTER_COMPONENT("doxygen" ENABLE_DOXYGEN DOXYGEN_FOUND)
+GR_REGISTER_COMPONENT("doxygen" ENABLE_DOXYGEN
```
We probably shouldn't be introducing this line break?

japm48 replied:
Oops, possibly a leftover of a previous discarded change.

marcusmueller commented on docs/doxygen/Doxyfile.swig_doc.in:
```diff
@@ -1318,7 +1311,7 @@ COMPACT_LATEX          = NO
 # executive. If left blank a4 will be used.
 
-PAPER_TYPE             = a4wide
+PAPER_TYPE             = letter
```
Why the paper format change?

drmpeg replied:
Doesn't matter, that SWIG file is now gone.

japm48 replied:
@marcusmueller I believe it introduced a warning, so I changed it to match with the regular `Doxyfile.in`. This file is no longer necessary as @drmpeg points out.

drmpeg replied:
@japm48 Just FYI, Marcus created a new PR and merged all this today.https://github.com/gnuradio/gnuradio/commit/2025e70872dcbfa0039ec000ea11d98d34d6cf70

marcusmueller commented:
> necromancer @marcusmueller to decide.

We prefer the term "demise-sceptics*, thank you.

Other than that: there's nothing to lose here. Sure, this is a new dependency, and MathJax itself isn't extremely fresh. But if it works in all has/hasn't mathjax/LaTeX combinations, then it just works.

Needs rebasing, and while we're at it, there's probably two changes in there that accidentally slipped in, so we might want to not make these two changes in the process of rebasing. Otherwise, this is good to go.

marcusmueller commented:
Continued in #3895
